linked_list/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+}
+
+// Связанный список
+#[derive(Debug)]
+pub struct LinkedList<T> {
+    pub head: Link<T>,
+    tail: Link<T>,
+}
+
+impl<T> LinkedList<T> {
+    fn new() -> Self {
+        LinkedList {
+            head: None,
+            tail: None,
+        }
+    }
+
+    // Добавление элемента в начало
+    pub fn push_front(&mut self, value: T) {
+        let new_node = Rc::new(RefCell::new(Node {
+            value,
+            next: self.head.clone(),
+        }));
+
+        self.head = Some(new_node.clone());
+        if self.tail.is_none() {
+            self.tail = Some(new_node);
+        }
+    }
+
+    // Добавление элемента в конец
+    pub fn push_back(&mut self, value: T) {
+        let new_node = Rc::new(RefCell::new(Node { value, next: None }));
+
+        match self.tail.take() {
+            Some(old_tail) => {
+                old_tail.borrow_mut().next = Some(new_node.clone());
+            }
+            None => {
+                self.head = Some(new_node.clone());
+            }
+        }
+        self.tail = Some(new_node);
+    }
+
+    // Добавление элемента после N-го
+    pub fn insert_after(&mut self, index: usize, value: T) {
+        let mut current = self.head.clone();
+        for _ in 0..index {
+            match current {
+                Some(ref node) => {
+                    let next = node.borrow().next.clone();
+                    current = next;
+                }
+                None => return,
+            }
+        }
+
+        if let Some(ref node) = current {
+            let new_node = Rc::new(RefCell::new(Node {
+                value,
+                next: node.borrow().next.clone(),
+            }));
+            let mut node_mut = node.borrow_mut();
+            if node_mut.next.is_none() {
+                self.tail = Some(new_node.clone());
+            }
+            node_mut.next = Some(new_node);
+        }
+    }
+
+    // Итератор по всем элементам
+    pub fn iter(&self) -> LinkedListIterator<T> {
+        LinkedListIterator {
+            current: self.head.clone(),
+        }
+    }
+
+    // Разделение списка на два списка
+    pub fn split_at(&mut self, index: usize) -> LinkedList<T> {
+        let mut current = self.head.clone();
+        let mut prev = None;
+        for _ in 0..index {
+            match current {
+                Some(ref node) => {
+                    prev = Some(node.clone());
+                    let next = node.borrow().next.clone();
+                    current = next;
+                }
+                None => return LinkedList::new(),
+            }
+        }
+
+        if let Some(ref node) = prev {
+            node.borrow_mut().next = None;
+        }
+
+        LinkedList {
+            head: current,
+            tail: self.tail.take(),
+        }
+    }
+}
+
+// Итератор для связанного списка
+pub struct LinkedListIterator<T> {
+    current: Link<T>,
+}
+
+impl<T> Iterator for LinkedListIterator<T> {
+    type Item = Rc<RefCell<Node<T>>>;
```

Comment:
Вот здесь вот должно быть вот так, чтобы не нагружать пользователя информацией о том, как внутри реализован список

```suggestion
    type Item = &T;
```

Reply:
Скачал ветку и попробовал сам поиграться. Давай забъем на это для простоты.
Да, просто так ссылку не вернуть, ей еще лайфтайм придется проставить, а для этого мы должны от чего-нибудь его взять. Обычно итераторы хранят ссылку на оригинальную коллекцию, откуда и вылезает этот лайфтайм. Но тут мы пользуемся счетчиком ссылок, так что давай так оставим